src/aiida_quantumespresso/workflows/functions/get_xspectra_structures.py

```diff
@@ -0,0 +1,325 @@
+                equivalency_dict[f'site_{key}'] = {}
+                atom_no_set = eq_atoms_data[key]
+                equivalency_dict[f'site_{key}']['site_index'] = key
+                equivalency_dict[f'site_{key}']['equivalent_sites_list'] = atom_no_set
+                equivalency_dict[f'site_{key}']['multiplicity'] = len(atom_no_set)
+                equivalency_dict[f'site_{key}']['symbol'] = site_symbol
+
+        output_params['equivalent_sites_data'] = equivalency_dict
+
+        x_pos = []
+        y_pos = []
+        z_pos = []
+        for site in structure.sites:
+            x_pos.append(site.position[0])
+            y_pos.append(site.position[1])
+            z_pos.append(site.position[2])
+
+        # We assume that the Martyna-Tuckerman approach will be used in the core-hole SCF
+        # thus, at a minimum, the new box for the molecule needs to be 2x the extent of the
+        # molecule in x, y, and z. We therefore use the larger of either the MT distance or
+        # `supercell_min_parameter` for each cell parameter.
+        high_x = x_pos[(np.argmax(x_pos))]
+        high_y = y_pos[(np.argmax(y_pos))]
+        high_z = z_pos[(np.argmax(z_pos))]
+
+        low_x = x_pos[(np.argmin(x_pos))]
+        low_y = y_pos[(np.argmin(y_pos))]
+        low_z = z_pos[(np.argmin(z_pos))]
+
+        box_a = max((high_x - low_x) * 2, supercell_min_parameter)
+        box_b = max((high_y - low_y) * 2, supercell_min_parameter)
+        box_c = max((high_z - low_z) * 2, supercell_min_parameter)
+
+        centred_structure = StructureData()
+        centred_structure.set_cell(value=([box_a, 0., 0.], [0., box_b, 0.], [0., 0., box_c]))
+
+        for kind in structure.kinds:
+            centred_structure.append_kind(kind)
+
+        for site in structure.sites:
+            centred_structure.append_site(site)
+
+        # Reset the positions so that the centre of mass is at the centre of the new box
+        centred_structure.reset_sites_positions(centred_positions + (np.array(centred_structure.cell_lengths) / 2))
+
+        output_params['centred_structure_cell_matrix'] = centred_structure.cell
+        output_params['centred_structure_cell_lengths'] = centred_structure.cell_lengths
+        result['centred_structure'] = centred_structure
+
+        for value in equivalency_dict.values():
+            target_site = value['site_index']
+            marked_structure = StructureData(pbc=False)
+            centred_kinds = {kind.name: kind for kind in centred_structure.kinds}
+
+            for index, site in enumerate(centred_structure.sites):
+                if index == target_site:
+                    absorbing_kind = Kind(name=abs_atom_marker, symbols=site.kind_name)
+                    absorbing_site = Site(kind_name=absorbing_kind.name, position=site.position)
+                    marked_structure.append_kind(absorbing_kind)
+                    marked_structure.append_site(absorbing_site)
+                else:
```

Comment:
You use `centered_structure` for the molecule. I suggest you use `supercell`, the same name as for the crystal. Here, the main goal is to create a larger unit cell for molecules, and shifting molecules to the center is only for better visualization, and will not affect the calculation (correct me if I am wrong).
If you use `supercell` for both of them, after creating the supercell, the step to generate the marked structures should be the same for molecule and crystal, thus lines 196-215 and lines 297-320 are the same and can be written only once.